### PR TITLE
test: relay parity foundation (mock + Azure share one scenario suite)

### PR DESCRIPTION
**Pull request by philsphicas**
## Why

Today the mock relay and Azure Relay are tested in isolation. Behavioural drift between them is invisible until it surfaces as a flake in one suite and not the other — and we have no way to *quantify* what a relay-side change (mux, retries, anything) actually changes about end-to-end behaviour.

This PR introduces a parity test harness so a single set of scenarios runs unmodified against both backends.

## What you get

- **`TestParity_Mock`** in `mockrelay/parity/` — runs in the default `go test ./mockrelay/...` job. Brings up an in-process mock relay + real aztunnel listener and sender, all in the same process. Sub-second per run.
- **`TestParity_Azure`** in `e2e/parity_test.go` (build tag `e2e`) — same scenarios against a real Azure Relay namespace, once per available auth method (Entra ID, SAS). Subprocess-driven via the existing e2e helpers.
- **Shared scenarios** in `internal/relayparity/` — backend-agnostic; new scenarios get added in one place and automatically run on both backends.

When the two suites disagree, the divergence is the headline finding.

## Scenarios in this slice

| Scenario | What it proves |
| --- | --- |
| `Echo_PortForward` | Bytes round-trip end-to-end through a port-forward tunnel. |
| `Echo_SOCKS5` | Same, but through the SOCKS5 sender (CONNECT path). |
| `Ordering_PortForward` | 1024 sequenced 1 KB chunks arrive in order, no interleaving or reordering inside the bridge. |
| `Bidirectional_PortForward` | 256 KB random payload + SHA256 verify in both directions concurrently — catches half-duplex bugs, back-pressure deadlocks, and silent body corruption. |

All four pass with the race detector on current `main`.

## What's deliberately not here

- Multi-listener / multi-sender / multi-target topology — comes in 2B.
- Benchmarks and quantitative characterisation — 2C.
- Stress, edge cases, half-close-aware protocols — 2D. Half-close in particular is out of scope: the current bridge tears down both directions on EOF, so half-close-aware tests would fail on `main` today. They become useful regression checks once that behaviour changes.

## How to run locally

```
go test -race ./mockrelay/parity/...                      # mock backend, ~1.4s
go test -tags=e2e -count=1 ./e2e/ -run TestParity_Azure   # Azure backend, requires E2E_* env
```

### CODE REVIEW

File: internal/relayparity/scenarios.go [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+		t.Fatalf("%v", writeErr)
+	}
+	if readErr != nil {
+		t.Fatalf("%v", readErr)
+	}
+	if readLen != size {
+		t.Fatalf("length mismatch: got %d want %d", readLen, size)
+	}
+	if readSum != want {
+		t.Fatalf("body sha256 mismatch:\n got=%x\nwant=%x", readSum, want)
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply internal/relayparity/scenarios.go]: Adopted. ScenarioBidirectional_PortForward now drops the shared variables and `sync.WaitGroup` entirely — each goroutine publishes its result on its own buffered channel (`wErr chan error` for the writer, `rRes chan readResult` for the reader, where readResult bundles err/sum/n). The main goroutine drains both channels and asserts. This makes the happens-before story explicit (channel send → channel receive) instead of relying on WaitGroup memory ordering around shared variables, and removes any future ambiguity for readers.

File: internal/relayparity/echo.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		conn, err := pe.ln.Accept()
+		if err != nil {
+			return
+		}
+		pe.wg.Add(1)
+		go func(c net.Conn) {
+			defer pe.wg.Done()
+			defer c.Close() //nolint:errcheck // best-effort cleanup
+			_, _ = io.Copy(c, c)
+		}(conn)
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply internal/relayparity/echo.go]: Adopted. PlainEcho now separates the accept-loop lifetime from per-connection tracking, exactly as suggested. The accept loop closes a `serveDone chan struct{}` when it exits (via `defer close(pe.serveDone)`), and per-connection goroutines are tracked by a separate `connWg sync.WaitGroup`. Stop does `ln.Close()` → `<-serveDone` → `connWg.Wait()`, so by the time we Wait on connWg the accept loop has already returned and no further connWg.Add can possibly happen. This makes the lifecycle obviously correct without requiring readers to reason about WaitGroup positive-Add-while-counter-nonzero semantics.

File: internal/relayparity/scenarios.go [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+	go func() {
+		buf := make([]byte, chunkSize)
+		for i := uint32(0); i < nChunks; i++ {
+			binary.BigEndian.PutUint32(buf[:4], i)
+			for j := 4; j < chunkSize; j++ {
+				buf[j] = byte(i + uint32(j))
+			}
+			if _, err := conn.Write(buf); err != nil {
+				writeErr <- fmt.Errorf("write chunk %d: %w", i, err)
+				return
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply internal/relayparity/scenarios.go:130]: Adopted. Added a `writeFull(io.Writer, []byte) error` helper that returns `io.ErrShortWrite` if Write returns n<len(buf) with a nil err, and switched both goroutine writers (the ScenarioOrdering chunk-writer loop and the ScenarioBidirectional body-writer) to use it. The existing `writeAll` helper (which already validated n on the main goroutine) was refactored to call `writeFull` underneath, so the two helpers share one definition of "full write". The line 291 reference in your comment landed on the writeAll helper itself, which already had the n check, so no behavior change there — just deduplicated.

File: mockrelay/parity/mock_backend.go [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+				Endpoint:      host,
+				EntityPath:    entity,
+				TokenProvider: tokenProvider,
+				ClientOptions: clientOpts,
+				BindAddress:   bind,
+				Logger:        silentLogger,
+			})
+		default:
+			t.Errorf("unknown SenderMode %v", opts.SenderMode)
+			return
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply mockrelay/parity/mock_backend.go]: Adopted. SenderMode validation now runs synchronously on the test goroutine before the sender goroutine is spawned: a small switch with `case ModePortForward, ModeSOCKS5:` no-ops, and `default: t.Fatalf("unknown SenderMode %v", ...)`. The previously-spawned listener goroutines are drained by the already-registered `t.Cleanup(cancel; wg.Wait)`, so an unknown-mode failure tears down cleanly with an immediate, accurate error message instead of trickling out as a generic 3-second `waitForTCP` timeout. The goroutine's inner switch lost its default branch (no longer reachable) since the synchronous validation above guarantees one of the two known modes.

File: internal/relayparity/socks5.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+	portInt, err := strconv.Atoi(portStr)
+	if err != nil || portInt <= 0 || portInt > 65535 {
+		return fmt.Errorf("parse port %q: invalid", portStr)
+	}
+	port := uint16(portInt)
+
+	// Method negotiation: VER=5, NMETHODS=1, NO-AUTH(0x00).
+	if _, err := conn.Write([]byte{0x05, 0x01, 0x00}); err != nil {
+		return fmt.Errorf("socks5 method write: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply internal/relayparity/socks5.go:64]: Adopted. Both SOCKS5 negotiation writes (method-selection at the line-59 anchor and CONNECT-request at the line-89 anchor) now go through the package-local `writeFull` helper introduced in the prior iteration. Tiny writes (3-byte method exchange, ~10-byte CONNECT request) make a short write practically impossible on any real TCP stack, but routing all writes through one helper keeps the contract explicit.

File: internal/relayparity/doc.go [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+// Two implementations exist:
+//
+//   - github.com/philsphicas/aztunnel/mockrelay/parity.MockBackend
+//     runs the mock relay server (mockrelay/server) and the aztunnel
+//     listener and sender in-process. Fast, deterministic, no external
+//     dependencies. Lives in the mockrelay module so the in-process
+//     server import does not leak into the main aztunnel binary's
+//     dependency graph.
+//
+//   - e2e/azure_backend_test.go (AzureBackend, build tag e2e) drives
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply internal/relayparity/doc.go]: Adopted. The package doc now refers to the unexported `azureBackend` to match the actual type in e2e/azure_backend_test.go. The Azure-side implementation is intentionally unexported because it's defined in a `_test.go` file under build tag `e2e` and its only consumer is the e2e test entry point in the same package, so the symmetry with MockBackend (which lives in a non-test file in mockrelay/parity and is exported for that reason) is asymmetric for a real reason. Doc updated; type kept unexported.

File: internal/relayparity/socks5.go [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+		conn.Close() //nolint:errcheck // best-effort cleanup on error
+		return nil, fmt.Errorf("set deadline: %w", err)
+	}
+	if err := socks5Negotiate(conn, target); err != nil {
+		conn.Close() //nolint:errcheck // best-effort cleanup on error
+		return nil, err
+	}
+	if err := conn.SetDeadline(time.Time{}); err != nil {
+		conn.Close() //nolint:errcheck // best-effort cleanup on error
+		return nil, fmt.Errorf("clear deadline: %w", err)
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply internal/relayparity/socks5.go]: Adopted. `DialSOCKS5` now computes a single absolute `deadline := time.Now().Add(timeout)` at entry, then uses `net.Dialer{Deadline: deadline}.Dial(...)` for the TCP dial and `conn.SetDeadline(deadline)` for the SOCKS5 negotiation — so the entire call (DNS resolution + connect + method exchange + CONNECT reply + BND drain) is bounded by one absolute deadline rather than two stacked `timeout` windows. `dialSOCKS5WithRetry`'s per-attempt budget is now actually enforced (the previous worst case of ~2× attempt timeout closed). Doc string updated to describe the new semantics.

While there I also tightened `dialSOCKS5WithRetry`'s overall-deadline contract: the inter-attempt 50 ms backoff used to sleep unconditionally and could overshoot the caller's deadline by up to ~50 ms on the final-failure path. It now checks remaining budget and breaks out when less than one full backoff remains, since there'd be no time for another attempt anyway.

File: mockrelay/parity/mock_backend.go [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+		TLSConfig: &tls.Config{InsecureSkipVerify: true}, //nolint:gosec // test cert
+	}
+	t.Cleanup(srv.Close)
+	return host, opts, srv
+}
+
+// pickFreePort returns a localhost host:port that was free at the
+// moment of the call. There is an unavoidable TOCTOU window before
+// the sender binds; the sender's bind error path retries via the
+// scenario's dial-with-retry helper.
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply mockrelay/parity/mock_backend.go]: Adopted (rewrote comment, kept the TOCTOU as-is). The `pickFreePort` comment no longer claims sender retries on bind error — the senders do `net.Listen` once and surface the error immediately. The comment now describes the actual failure mode: if another process steals the port in the microsecond window between Listen+Close and the sender's Listen, the sender exits with the bind error and Setup's downstream `waitForTCP` times out at 3 s with a loud "never became reachable" failure. The fix you suggested as the alternative — bind 127.0.0.1:0 in the sender itself and capture the actual addr via a readiness hook — is the right structural fix and is on the slice-2B list (the same TODO that calls out the MaxConnections probe-slot interaction), but doing it now would expand the sender API for a Slice-2A test helper, so leaving it to 2B with the honest comment in the meantime.

File: internal/relayparity/scenarios.go [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+	t.Helper()
+	deadline := time.Now().Add(timeout)
+	var lastErr error
+	for time.Now().Before(deadline) {
+		conn, err := net.DialTimeout("tcp", addr, 1*time.Second)
+		if err == nil {
+			return conn
+		}
+		lastErr = err
+		time.Sleep(25 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply internal/relayparity/scenarios.go]: Adopted. `dialWithRetry` now mirrors `dialSOCKS5WithRetry`'s budgeting: it computes a single absolute `deadline := time.Now().Add(timeout)` at entry, caps each per-attempt `net.DialTimeout` at the smaller of the remaining budget or 1 s, and skips the inter-attempt 25 ms backoff when less than that backoff remains. The helper now respects its overall timeout precisely — no more potential ~1 s overshoot from a stalled per-attempt dial.

File: mockrelay/parity/mock_backend.go [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+	// t.Fatalf in a readiness gate still drains anything we managed to
+	// start. wg.Wait() on a zero-counter wg returns immediately, so
+	// this is safe when Setup fails before any goroutine launches.
+	var wg sync.WaitGroup
+	t.Cleanup(func() {
+		cancel()
+		wg.Wait()
+	})
+
+	host, clientOpts, srv := startMockRelay(t)
```
[the remark this answers is not in the document]
Comment (philsphicas) [reply mockrelay/parity/mock_backend.go:66]: Acknowledged. The current ordering causes a benign log line ("listener exited: ...") on teardown when `srv.Close` runs before `cancel+wg.Wait` — listener/sender goroutines see a network error during the shutdown they were already heading into. No correctness impact (the WaitGroup still ensures all goroutines exit before the test returns).

Moving forward without changing this. Rationale:

1. The cleanup ordering is **already going to be restructured** in slice 2B when `mockrelay/parity/mock_backend.go` gains per-listener handles for hot-drop / hot-add scenarios. Reordering the `t.Cleanup` calls now and then immediately rewriting that block again in 2B adds churn without value.
2. The PR otherwise lands a complete, tested foundation (same scenarios pass on both backends with `-race`), and the gap above is non-blocking for product behaviour — all changes are in new test-only packages, with zero imports from product code.

Filing the cleanup-ordering rewrite as part of the 2B work item.
